File: src-py/src/subsystems/drive.py

from wpilib.command.subsystem import Subsystem

from wpilib import RobotDrive

from hardware.motor import Motor
from values import ids

from wpilib.solenoid import Solenoid

class Drive(Subsystem):

	def __init__(self):
		super().__init__('Drive')

		# These can be used individually, but you should use RobotDrive to controll these
		self.LB = Motor(ids.LB_motor)
		self.LF = Motor(ids.LF_motor)
		self.RB = Motor(ids.RB_motor)
		self.RF = Motor(ids.RF_motor)

		self.LB.setInverted(True)
		self.LF.setInverted(True)

		self.drive_train = RobotDrive(self.LF, self.LB, self.RF, self.RB)
		self.drive_train.setExpiration(0.1)

		## Uncomment to fail
		#import time
		#time.sleep(5)
		# doesn't work bc HAL Resource error
		#self.pcm0 = Solenoid(ids.PCM_pin0)  
		#self.pcm1 = Solenoid(ids.PCM_pin1)

	def stop(self):
		self.tank(0, 0)

	def tank(self, L_speed, R_speed):
		self.drive_train.tankDrive(L_speed, R_speed, squaredInputs=False)

	def mecanum_cartesian(self, X_speed, Y_speed, R_speed=0, gyroHeading=0):
		self.drive_train.mecanumDrive_Cartesian(X_speed, Y_speed, R_speed, gyroHeading)

	def mecanum_polar(self, _speed, _direction, R_speed=0):
		self.drive_train.mecanumDrive_Polar(_speed, _direction, R_speed)

File: src-java/src/main/java/frc3966/robot/subsystems/Drive.java
package frc3966.robot.subsystems;

import frc3966.robot.hardware.DriveMotor;
import edu.wpi.first.wpilibj.command.Subsystem;
import frc3966.robot.commands.MecDrive;
import frc3966.robot.values.IDs;
import edu.wpi.first.wpilibj.RobotDrive;

/**
 * Drive subsystem, controls left and right motors.
 *
 * Supplies methods for controlling the drive by various functions. The motors
 * objects should not be directly set from any other place.
 */
public class Drive extends Subsystem {

	public Drive() {
		LF = new DriveMotor(IDs.LF_motor);
		LB = new DriveMotor(IDs.LB_motor);
		RF = new DriveMotor(IDs.RF_motor);
		RB = new DriveMotor(IDs.RB_motor);
		LB.setInverted(true);
		LF.setInverted(true);
		robotDrive = new RobotDrive(LF, LB, RF, RB);
	}

	private DriveMotor LB, LF, RB, RF;
	private RobotDrive robotDrive;

	public void initDefaultCommand() {
		setDefaultCommand(new MecDrive());
	}

	public void stop() {
		tank(0, 0);
	}
	
	public void tank(double L_speed, double R_speed) {
		robotDrive.tankDrive(L_speed, R_speed);
	}
	
	public void mecanum_cartesian(double X_speed, double Y_speed, double R_speed) {
		robotDrive.mecanumDrive_Cartesian(X_speed, Y_speed, R_speed, 0);
	}
	public void mecanum_polar(double _speed, double _direction, double R_speed) {
		robotDrive.mecanumDrive_Polar(_speed, _direction, R_speed);
	}

}
